File: commit-tree.go
// cat-file.go

package main

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"os/user"
	"strings"
	"time"
)

const ORIG_OFFSET = 40

// -----------------------------------------------------------------------------------------------

func main() {
	if len(os.Args) < 2 {
		usage("commit-tree <sha1> [-p <sha1>]* < changelog")
	}

	// Turn the hex string into a Sha1 value
	tree_sha1, err := get_sha1_hex(os.Args[1])
	if err != nil {
		usage("commit-tree <sha1> [-p <sha1>]* < changelog")
	}

	// Get any extra parents
	var parents [][20]byte
	for i := 2; i < len(os.Args); i = i + 2 {
		if os.Args[i] == "-p" {
			sha1, err := get_sha1_hex(os.Args[i+1])
			if err == nil {
				parents = append(parents, sha1)
				continue
			}
		}
		usage("commit-tree <sha1> [-p <sha1>]* < changelog")
	}

	if len(parents) == 0 {
		fmt.Fprintf(os.Stderr, "Committing initial tree %s\n", os.Args[1])
	}

	// For committer, we use machine name+email and current time
	// For author, we use env vars if they exist, otherwise author info
	// The idea is you might have a patch from someone else, but you always
	// commit as your local self.

	// Get login name and email to use for committer info
	var committer_name, committer_email string
	if committer_name, committer_email = getUserEmail(); committer_name == "" {
		usage("You don't exist. Go away!")
	}

	// Get current datetime in Git-desired format
	committer_date := time.Now().Format("Mon Jan _2 15:04:05 2006")

	// Override machine values with environment variables if they exist
	// Note that these environment variables are named backwards. I didn't do it.
	var author_name, author_email, author_date string
	if author_name = os.Getenv("COMITTER_NAME"); author_name == "" {
		author_name = committer_name
	}
	if author_email = os.Getenv("COMMITTER_EMAIL"); author_email == "" {
		author_email = committer_email
	}
	if author_date = os.Getenv("COMMITTER_DATE"); author_date == "" {
		author_date = committer_date
	}

	// Fix strings, removing garbage from OS
	committer_name = strings.Replace(committer_name, "[]\n", "", -1)
	committer_email = strings.Replace(committer_email, "[]\n", "", -1)
	committer_date = strings.Replace(committer_date, "[]\n", "", -1)

	author_name = strings.Replace(author_name, "[]\n", "", -1)
	author_email = strings.Replace(author_email, "[]\n", "", -1)
	author_date = strings.Replace(author_date, "[]\n", "", -1)

	// Create a new buffer to hold a tree, and create a blank
	// spacer that we will fill in with the object header once we
	// know the size of the tree object.
	var b bytes.Buffer
	fmt.Fprintf(&b, "%*s", ORIG_OFFSET, "")

	// Add first parent
	fmt.Fprintf(&b, "tree %s\n", sha1_to_hex(tree_sha1))

	// Add any remaining parents
	for _, parent := range parents {
		fmt.Fprintf(&b, "parent %s\n", sha1_to_hex(parent))
	}

	// Add author and committer info
	fmt.Fprintf(&b, "author %s <%s> %s\n", author_name, author_email, author_date)
	fmt.Fprintf(&b, "committer %s <%s> %s\n", committer_name, committer_email, committer_date)

	// Add the comment from stdin
	io.Copy(&b, os.Stdin)

	// Finalize by adding a header. Because the header is not fixed-length,
	// we need to skip past any pad bytes when we write it out.
	buf := b.Bytes()
	offset := finish_buffer("commit", buf)
	write_sha1_file(buf[offset:])
}

// finish_buffer finalizes the buffer by adding an object header.
// It returns the offset needed to skip past unneeded starting pad bytes
func finish_buffer(tag string, buf []byte) int {
	tagline := fmt.Sprintf("%s %d\x00", tag, len(buf[40:]))

	// copy it into the buffer so it aligns with the start of object data
	offset := ORIG_OFFSET - len(tagline)
	copy(buf[offset:], tagline)

	// This is the number of unused bytes at the beginning of the buffer
	return offset
}

// -----------------------------------------------------------------------------------------------

// Go doesn't really have this in the format that getpwuid
// provides, so create it from what we have available
func getUserEmail() (name, email string) {

	// Do our best to get a real username; if we can't, use
	// the login name
	var username, login string
	if me, err := user.Current(); err == nil {
		login = me.Username
		if username = me.Name; me.Name == "" {
			username = login
		}
	}

	// Get hostname. If we can't get a hostname, we are out of luck
	var err error
	var hostname string
	if hostname, err = os.Hostname(); err != nil {
		return "", ""
	}

	// On Windows, the username usually is prefixed by the hostname,
	// so strip it
	if len(username) > len(hostname) && username[0:len(hostname)] == hostname {
		username = username[len(hostname):]
		if username[0] == '\\' {
			username = username[1:]
		}
	}

	// Note that "user@host" is unlikely to be a valid email address
	// on Windows, but it's the best we can do
	return username, fmt.Sprintf("%s@%s", login, hostname)
}

File: commit-tree.c
#include "cache.h"

#include <time.h>

#define BLOCKING (1ul << 14)
#define ORIG_OFFSET (40)

/*
 * Leave space at the beginning to insert the tag
 * once we know how big things are.
 *
 * FIXME! Share the code with "write-tree.c"
 */
static void init_buffer(char **bufp, unsigned int *sizep)
{
	char *buf = malloc(BLOCKING);
	memset(buf, 0, ORIG_OFFSET);
	*sizep = ORIG_OFFSET;
	*bufp = buf;
}

static void add_buffer(char **bufp, unsigned int *sizep, const char *fmt, ...)
{
	char one_line[2048];
	va_list args;
	int len;
	unsigned long alloc, size, newsize;
	char *buf;

	va_start(args, fmt);
	len = vsnprintf(one_line, sizeof(one_line), fmt, args);
	va_end(args);
	size = *sizep;
	newsize = size + len;
	alloc = (size + 32767) & ~32767;
	buf = *bufp;
	if (newsize > alloc) {
		alloc = (newsize + 32767) & ~32767;
		buf = realloc(buf, alloc);
		*bufp = buf;
	}
	*sizep = newsize;
	memcpy(buf + size, one_line, len);
}

static int prepend_integer(char *buffer, unsigned val, int i)
{
	buffer[--i] = '\0';
	do {
		buffer[--i] = '0' + (val % 10);
		val /= 10;
	} while (val);
	return i;
}

static void finish_buffer(char *tag, char **bufp, unsigned int *sizep)
{
	int taglen;
	int offset;
	char *buf = *bufp;
	unsigned int size = *sizep;

	offset = prepend_integer(buf, size - ORIG_OFFSET, ORIG_OFFSET);
	taglen = strlen(tag);
	offset -= taglen;
	buf += offset;
	size -= offset;
	memcpy(buf, tag, taglen);

	*bufp = buf;
	*sizep = size;
}

static void remove_special(char *p)
{
	char c;
	char *dst = p;

	for (;;) {
		c = *p;
		p++;
		switch(c) {
		case '\n': case '<': case '>':
			continue;
		}
		*dst++ = c;
		if (!c)
			break;
	}
}

/*
 * Having more than two parents may be strange, but hey, there's
 * no conceptual reason why the file format couldn't accept multi-way
 * merges. It might be the "union" of several packages, for example.
 *
 * I don't really expect that to happen, but this is here to make
 * it clear that _conceptually_ it's ok..
 */
#define MAXPARENT (16)

// Struct containing bits of commit info that are fetched in a machine-specific way
struct commit_info
{
    char* gecos;
    char* email;
    char* date;
    char* realgecos;
    char realemail[1000];
    char* realdate;
    char comment[1000];
};
int fill_commit_info(struct commit_info* info); // per-platform routine

int main(int argc, char **argv)
{
	int i;
	int parents = 0;
	unsigned char tree_sha1[20];
	unsigned char parent_sha1[MAXPARENT][20];
    struct commit_info info;
	char *buffer;
	unsigned int size;

	if (argc < 2 || get_sha1_hex(argv[1], tree_sha1) < 0)
		usage("commit-tree <sha1> [-p <sha1>]* < changelog");

	for (i = 2; i < argc; i += 2) {
		char *a, *b;
		a = argv[i]; b = argv[i+1];
		if (!b || strcmp(a, "-p") || get_sha1_hex(b, parent_sha1[parents]))
			usage("commit-tree <sha1> [-p <sha1>]* < changelog");
		parents++;
	}
	if (!parents)
		fprintf(stderr, "Committing initial tree %s\n", argv[1]);

    if (!fill_commit_info(&info))
		usage("You don't exist. Go away!");

	init_buffer(&buffer, &size);
	add_buffer(&buffer, &size, "tree %s\n", sha1_to_hex(tree_sha1));

	/*
	 * NOTE! This ordering means that the same exact tree merged with a
	 * different order of parents will be a _different_ changeset even
	 * if everything else stays the same.
	 */
	for (i = 0; i < parents; i++)
		add_buffer(&buffer, &size, "parent %s\n", sha1_to_hex(parent_sha1[i]));

	/* Person/date information */
	add_buffer(&buffer, &size, "author %s <%s> %s\n", info.gecos, info.email, info.date);
	add_buffer(&buffer, &size, "committer %s <%s> %s\n\n", info.realgecos, info.realemail, info.realdate);

	/* And add the comment */
	while (fgets(info.comment, sizeof(info.comment), stdin) != NULL)
		add_buffer(&buffer, &size, "%s", info.comment);

	finish_buffer("commit ", &buffer, &size);

	write_sha1_file(buffer, size);
	return 0;
}

#ifdef __linux__
#include <pwd.h>
int fill_commit_info(struct commit_info* info)
{
    int len;
	struct passwd *pw;
	time_t now;

	pw = getpwuid(getuid());
	if (!pw)
		return 0;
	info->realgecos = pw->pw_gecos;
	len = strlen(pw->pw_name);
	memcpy(info->realemail, pw->pw_name, len);
	info->realemail[len] = '@';
	gethostname(info->realemail+len+1, sizeof(info->realemail)-len-1);
	time(&now);
	info->realdate = ctime(&now);

	info->gecos = getenv("COMMITTER_NAME"); if (!info->gecos) info->gecos = info->realgecos;
	info->email = getenv("COMMITTER_EMAIL"); if (!info->email) info->email = info->realemail;
	info->date = getenv("COMMITTER_DATE"); if (!info->date) info->date = info->realdate;

	remove_special(info->gecos); remove_special(info->realgecos);
	remove_special(info->email); remove_special(info->realemail);
	remove_special(info->date); remove_special(info->realdate);

    return 1;
}
#endif
